init.lua

```diff
@@ -1,116 +1,540 @@
-local original_pos = {}
+-- NOTE: in the output texts, the names are always in double quotes because some players have
+--	names that can be confusing without the quotes.
+-- IDEA: technically it would be possible to chain observe. Would have to climb the parent tree
+--	making sure there is nothing circular happening. Including checking all the children.
+--	A lot can go wrong with that, so it has been left out for now.
+--	Another complication to this is that there are many combinations of client<->server
+--	software versions to consider.
+-- IDEA: might be nice to have a /send_home (<player>|all) command for invitie to detach
+--	invited guests again.
+--	Currently player can force detachment by logging off.
+spectator_mode = {
+	version = 20220214,
+	command_accept = minetest.settings:get('spectator_mode.command_accept') or 'smy',
+	command_deny = minetest.settings:get('spectator_mode.command_deny') or 'smn',
+	command_detach = minetest.settings:get('spectator_mode.command_detach') or 'unwatch',
+	command_invite = minetest.settings:get('spectator_mode.command_invite') or 'watchme',
+	command_attach = minetest.settings:get('spectator_mode.command_attach') or 'watch',
+	invitation_timeout = tonumber(minetest.settings:get(
+		'spectator_mode.invitation_timeout') or 1 * 60),
+
+	keep_all_observers_alive = minetest.settings:get_bool(
+		'spectator_mode.keep_all_observers_alive', false),
+
+	priv_invite = minetest.settings:get('spectator_mode.priv_invite') or 'interact',
+	priv_watch = minetest.settings:get('spectator_mode.priv_watch') or 'watch',
+}
+local sm = spectator_mode
+do
+	local temp = minetest.settings:get('spectator_mode.extra_observe_privs') or ''
+	sm.extra_observe_privs, sm.extra_observe_privs_moderator = {}, nil
+	for _, priv in ipairs(temp:split(',')) do
+		sm.extra_observe_privs[priv] = true
+	end
+	temp = minetest.settings:get('spectator_mode.extra_observe_privs_moderator')
+	if (not temp) or ('' == temp) then
+		sm.extra_observe_privs_moderator = sm.extra_observe_privs
+	else
+		sm.extra_observe_privs_moderator = {}
+		for _, priv in ipairs(temp:split(',')) do
+			sm.extra_observe_privs_moderator[priv] = true
+		end
+	end
+end
+-- pull some global references to local space
+local after = minetest.after
+local chat = minetest.chat_send_player
+local core_log = minetest.log
+local deserialize = minetest.deserialize
+local serialize = minetest.serialize
+local get_player_privs = minetest.get_player_privs
+local set_player_privs = minetest.set_player_privs
+local get_player_by_name = minetest.get_player_by_name
+local vector_new = vector.new
+local vector_round = vector.round
```

Comment:
Do you have any performance statistics about local vs global variable access? This seems to be premature optimization.

Reply:
No, I have not profiled. Doesn't it add to readability too?

I'm sure you are familiar with http://www.lua.org/pil/4.2.html and discusions around it:
> A common idiom in Lua is
>
>    local foo = foo
> This code creates a local variable, foo, and initializes it with the value of the global variable foo. That idiom is useful when the chunk needs to preserve the original value of foo even if later some other function changes the value of the global foo; it also **speeds up access to foo**.

If anybody wants to performance test this, please do so from outside the running Lua environment.

Reply:
> That idiom is useful when the chunk needs to preserve the original value of foo even if later some other function changes the value of the global foo

This part is often unwanted with minetest mods, sometimes it is wanted but most often you want to allow other mods to change behavior of globals.
This exact set of selected globals are less likely to be overridden, not most common ones, but still situation is similar that you'd get for example with `local is_protected = minetest.is_protected`.

Reply:
changed in https://github.com/SwissalpS/spectator_mode/commit/0eca5efb5a79444d599aff57a8a374602c61f088